src/main/java/org/truffleruby/core/array/ArrayNodes.java

```diff
@@ -1581,6 +1581,62 @@ public DynamicObject replace(DynamicObject array, DynamicObject other,
 
     }
 
+    @Primitive(name = "array_rotate_inplace", needsSelf = false, lowerFixnum = 2)
+    @ImportStatic(ArrayGuards.class)
+    public abstract static class RotateInplaceNode extends PrimitiveArrayArgumentsNode {
+
+        @Specialization(guards = "strategy.matches(array)", limit = "ARRAY_STRATEGIES")
+        public DynamicObject rotate(DynamicObject array, int amount,
+                @Cached("of(array)") ArrayStrategy strategy) {
+            final int size = strategy.getSize(array);
+            assert 0 < amount && amount < size;
+            final ArrayMirror mirror = strategy.newMirror(array);
+
+            rotateReverse(amount, size, mirror);
+
+            return array;
+        }
+
+        protected void rotateReverse(int amount, int size, ArrayMirror mirror) {
+            final int last = size - 1;
+            reverse(mirror, amount, last);
+            reverse(mirror, 0, amount - 1);
+            reverse(mirror, 0, last);
+        }
+
+        private void reverse(ArrayMirror mirror, int from, int to) {
```

Comment:
Would it make sense to `@ExlpodeLoop` for a cachedSize specialization?

Reply:
if `to - from` is very small, it could be interesting indeed.

Reply:
It gives around 2x on `ary.rotate!(8)` (ary has size 16) by also changing the logic to just connect load and stores (see 94288ce), thanks for the idea!
It's quite nice in IGV:
![screenshot from 2017-09-30 15-22-54](https://user-images.githubusercontent.com/168854/31046214-3c643b90-a5f5-11e7-9781-db9b7c767e03.png)

Reply:
👍